File: Core/C#/BinarySearch.cs
namespace C_
{
    public class BinarySearch
    {
        // Binary search, keep halving the array untill we find
        // that x is a middle point.
        // runtime = N(log n)
        public static int Search(int[] arr, int left, int right, int x)
        {
            if (right > left)
            {
                // We know we are searching normally.
                int middle = left + (r - 1) / 2;
                // If middle IS x then just return that.
                if (arr[middle] == x) return middle;

                // If element is smaller than mid, then can only be present in left array,
                // and vice versa.
                // (middle - 1 = the new right for the recursive call.)
                if (x < arr[middle]) return Search(arr, left, middle - 1, x);
                return Search(arr, middle + 1, right, x);
            }
            // If no item was returned the answer is not found.
            return -1;
        }
    }
}

File: Core/C++/Searching/binary-search.cpp
/*
    Performs a simple binary search on an array of integers.
*/

int binarySearch(int arr[], int value, int length)
{
    return binarySearchInner(arr, value, 0, length - 1);
}

int binarySearchInner(int arr[], int value, int left, int right)
{
    if (left > right) return -1;
    int middle = (right - left) / 2;
    if (arr[middle] == value) return middle;
    if (arr[left] == value) return left;
    if (arr[right] == value) return right;

    if (value < middle) return binarySearchInner(arr, value, left, middle - 1);
    return binarySearchInner(arr, value, middle + 1, right);
}